src/passes/Inlining.cpp

```diff
@@ -158,10 +184,56 @@ struct FunctionInfoScanner
     }
 
     info.size = Measurer::measure(curr->body);
+
+    info.maySkipHeavyWork = maySkipHeavyWork(curr->body);
   }
 
 private:
   NameInfoMap* infos;
+
+  // Check if a function may avoid doing any heavy work at all.
+  bool maySkipHeavyWork(Expression* curr) {
+    Index avoidableWork = 0;
+    Index unavoidableWork = 0;
+
+    auto process = [&](Expression* child) {
+      if (auto* iff = child->dynCast<If>()) {
+        // The condition cannot be avoided.
+        unavoidableWork += CostAnalyzer(iff->condition).cost;
+
+        // Only one arm will be taken.
+        Index ifTrue = CostAnalyzer(iff->ifTrue).cost;
+        Index ifFalse = iff->ifFalse ? CostAnalyzer(iff->ifFalse).cost : 0;
+        avoidableWork += std::max(ifTrue, ifFalse);
+        unavoidableWork += std::min(ifTrue, ifFalse);
+        return;
+      }
+
+      // If this is not a node we can handle, just assume the cost is
+      // unavoidable. TODO: add more node types
+      unavoidableWork += CostAnalyzer(child).cost;
+    };
+
+    // Look into a block's children; otherwise, just scan the node.
+    if (auto* block = curr->dynCast<Block>()) {
+      for (auto* child : block->list) {
+        process(child);
+      }
+    } else {
+      process(curr);
+    }
+
+    // Limit the unavoidable work to really trivial things, like say one or two
+    // eqz, ref.is_null, etc. operations.
+    const Index MaxUnavoidable = 2;
+
+    // To count as avoiding heavy work, the avoidable work must be pretty
+    // significant - something like a call, throw, etc. This should be something
+    // like an order of magnitude higher than MaxUnavoidable.
+    const Index MinAvoidable = 20;
+
+    return unavoidableWork <= MaxUnavoidable && avoidableWork >= MinAvoidable;
```

Comment:
If the avoidable work is huge, it probably isn't worth the code size hit of inlining. Should we have an upper bound on the avoidable work as well?

Reply:
This does not supersede the usual code size limits. This just avoids us dismissing an opportunity just because of seeing a call or a loop or something else heavy.

I was thinking that's enough, but now that you mention it maybe it's not. We could make this rule only apply when not optimizing for size - how does that sound?

Reply:
Ah that makes sense. Yes, using this only when not optimizing for size sounds right to me.

Reply:
Actually, it turns out that was already the case - just the code was a little unclear. I reordered it and added more comments.